Validate URL if validation data does not exist while generating support data.

dhaval-parekh opened the pull request:
## Summary

<!-- Please reference the issue(s) this PR fixes, if one exists. For significant changes, opening an issue first for discussion is usually preferable. -->
Fixes #6707 

This PR adds changes to support request flow. With new changes, while preparing support data if validation data for the requested URL doesn't found it will validate and store the data for the URL (which previously doesn't happen)

Also, Added appropriate messages when there is no AMP related error information.

1. When there is no validation data.
<img width="1110" alt="image" src="https://user-images.githubusercontent.com/8168027/142410945-8213d45b-f6f9-407a-a1b3-f0f37be0cda3.png">

2. When all validation data are stalled.
<img width="1098" alt="image" src="https://user-images.githubusercontent.com/8168027/142411883-3e992a39-1c37-4c3a-ad84-e3d52bc048a2.png">

3. When there is no AMP error on validated URLs but there is some stalled result.
<img width="1106" alt="image" src="https://user-images.githubusercontent.com/8168027/142412168-08164d1b-d0b4-4965-b4bf-e391de9cb062.png">

4. When there is no AMP error on all validated URLs.
<img width="1094" alt="image" src="https://user-images.githubusercontent.com/8168027/142411816-ba2e415d-69d0-4f60-99b7-8577b00eda4f.png">

5. When request URL doesn't have AMP error.
<img width="1096" alt="image" src="https://user-images.githubusercontent.com/8168027/142411378-073710b0-cf75-44f2-841d-59fa575931f2.png">


## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

dhaval-parekh committed:
Pass the front-end URL to support page instead of post id of amp_validated_url

dhaval-parekh committed:
Validate URL while sending support data if data not exist for URL

dhaval-parekh committed:
Update unit test cases

dhaval-parekh committed:
Refactor SupportData service.
Add notice in AMP Support page if there is not AMP error

dhaval-parekh committed:
Update unit test cases

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+				</AMPNotice>
+			) }
+		</>
+	);
```
Since all those notices are mutually exclusive (or maybe they aren't?) I think it'd be both easier to read and slightly more performant if we converted those code blocks into a series of regular `if` statements with `return` inside, like:

```js
if ( isSpecificUrlRequest && hasError ) {
	return null;
}

if ( isSpecificUrlRequest && ! hasErrors ) {
	return (
		<AMPNotice type={ NOTICE_TYPE_SUCCESS } size={ NOTICE_SIZE_SMALL }>
			...
		</AMPNotice>
	);
}

if ( ! isSpecificUrlRequest && 0 === ampValidatedPostCount.all ) {
	return (
		<AMPNotice type={ NOTICE_TYPE_WARNING } size={ NOTICE_SIZE_SMALL }>
			...
		</AMPNotice>
	);
}

[ ... ]
```

delawski commented on assets/src/components/amp-support/index.js:
```diff
@@ -94,6 +95,9 @@ export function AMPSupport( props ) {
 					}
 				} />
+
+				{ <Notices data={ data } args={ args } ampValidatedPostCount={ ampValidatedPostCount } /> }
```
There's no need to wrap the component in curly braces:

```suggestion
				<Notices data={ data } args={ args } ampValidatedPostCount={ ampValidatedPostCount } />
```

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+ * @return {JSX.Element|null} HTML markup for notice data.
+ */
+export function Notices( props ) {
+	const { data, args, ampValidatedPostCount } = props;
```
Would it make sense to name the component in a more specific way, e.g. `ValidationResultsNotice`?

Also, we're not using `props` anywhere in the component so we could destructure it directly inside the function declaration:

```suggestion
export function ValidationResultsNotice( { data, args, ampValidatedPostCount } ) {
```

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+	const { data, args, ampValidatedPostCount } = props;
+
+	const isSpecificUrlRequest = ( 0 < args?.urls?.length );
+	const hasErrors = ( 0 < data?.errors?.length || 0 !== data?.urls?.length );
```
1. Since `data.errors` is marked as `isRequired`, do we need the optional chaining here?
2. We could use the already determined `isSpecificUrlRequest` value here to make the code more readable.
3. There's no need for parentheses:

```suggestion
	const hasErrors = 0 < data.errors?.length || isSpecificUrlRequest;
```

Lastly, are we always sure that when the `args.urls` is non-empty then there are errors?

dhaval-parekh replied:
> 1. Since `data.errors` is marked as `isRequired`, do we need the optional chaining here?

No, We can remove chaining.

> 2.  We could use the already determined `isSpecificUrlRequest` value here to make the code more readable.

There is a difference in condition here. For `isSpecificUrlRequest` we are check URLs length of `args` and for `hasErrors` we need to check URLs the length of data.

> Lastly, are we always sure that when the `args.urls` is non-empty then there are errors?

No. That won't be always the case. non-empty value for `args.urls` means a user has requested to send support data for specific URLs.

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+export function Notices( props ) {
+	const { data, args, ampValidatedPostCount } = props;
+
+	const isSpecificUrlRequest = ( 0 < args?.urls?.length );
```
There's no need for parentheses:
```suggestion
	const isSpecificUrlRequest = 0 < args?.urls?.length;
```

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+
+	const isSpecificUrlRequest = ( 0 < args?.urls?.length );
+	const hasErrors = ( 0 < data?.errors?.length || 0 !== data?.urls?.length );
+	const hasAllStalledResult = ( 0 !== ampValidatedPostCount.all && 0 === ampValidatedPostCount.valid );
```
1. We should use the term `stale` instead of `stall`.
2. No need for parentheses.
```suggestion
	const hasAllStaleResults = 0 < ampValidatedPostCount.all && 0 === ampValidatedPostCount.valid;
```

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+	const isSpecificUrlRequest = ( 0 < args?.urls?.length );
+	const hasErrors = ( 0 < data?.errors?.length || 0 !== data?.urls?.length );
+	const hasAllStalledResult = ( 0 !== ampValidatedPostCount.all && 0 === ampValidatedPostCount.valid );
+	const hasSomeStalledResult = ( 0 !== ampValidatedPostCount.all && 0 !== ampValidatedPostCount.valid && ampValidatedPostCount.all !== ampValidatedPostCount.valid );
```
1. We should use the term `stale` instead of `stall`.
2. No need for parentheses.
3. I think we can get rid of one condition: if `valid` is greater than 0 then `all` must be greater than 0, too.
```suggestion
	const hasSomeStaleResults = 0 < ampValidatedPostCount.valid && ampValidatedPostCount.all !== ampValidatedPostCount.valid;
```

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+	const hasErrors = ( 0 < data?.errors?.length || 0 !== data?.urls?.length );
+	const hasAllStalledResult = ( 0 !== ampValidatedPostCount.all && 0 === ampValidatedPostCount.valid );
+	const hasSomeStalledResult = ( 0 !== ampValidatedPostCount.all && 0 !== ampValidatedPostCount.valid && ampValidatedPostCount.all !== ampValidatedPostCount.valid );
+	const hasAllValidResult = ( 0 !== ampValidatedPostCount.all && ampValidatedPostCount.all === ampValidatedPostCount.valid );
```
There's no need for adding parentheses:
```suggestion
	const hasAllValidResults = 0 < ampValidatedPostCount.all && ampValidatedPostCount.all === ampValidatedPostCount.valid;
```

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+			{ /* All validated URLs of site is stalled. */ }
+			{ ! isSpecificUrlRequest && hasAllStalledResult && (
+				<AMPNotice type={ NOTICE_TYPE_WARNING } size={ NOTICE_SIZE_SMALL }>
+					{ __( 'The validation data are stalled. Please re-scan your site from the "AMP Setting" page to send enough data for support requests.', 'amp' ) }
```
```suggestion
					{ __( 'The validation data is stale. Go to the AMP Settings page and rescan you site before sending a support request.', 'amp' ) }
```

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+			{ /* The site doesn't have any AMP errors but there are some stalled result. */ }
+			{ ! isSpecificUrlRequest && ! hasErrors && hasSomeStalledResult && (
+				<AMPNotice type={ NOTICE_TYPE_WARNING } size={ NOTICE_SIZE_SMALL }>
+					{ __( 'We found no issues on your site but there are some stalled result. Browse your site to ensure everything is working as expected.', 'amp' ) }
```
```suggestion
					{ __( 'We found no issues on your site but there are some stale validation results. Browse your site to ensure everything is working as expected.', 'amp' ) }
```

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+			{ /* If requested URL doesn't have AMP related errors. */ }
+			{ isSpecificUrlRequest && ! hasErrors && (
+				<AMPNotice type={ NOTICE_TYPE_SUCCESS } size={ NOTICE_SIZE_SMALL }>
+					{ __( 'The requested URL does not have any AMP related errors.', 'amp' ) }
```
```suggestion
					{ __( 'The requested URL does not have any AMP validation errors.', 'amp' ) }
```

delawski commented on assets/src/components/amp-support/notices.js:
```diff
@@ -0,0 +1,91 @@
+			{ /* The site doesn't have any validated URLs. */ }
+			{ ! isSpecificUrlRequest && 0 === ampValidatedPostCount.all && (
+				<AMPNotice type={ NOTICE_TYPE_WARNING } size={ NOTICE_SIZE_SMALL }>
+					{ __( 'The site has no validation data. Please scan your site from the "AMP Setting" page to send enough data for support requests.', 'amp' ) }
```
```suggestion
					{ __( 'The site has no validation data. Go to the AMP Settings page and scan you site before sending a support request.', 'amp' ) }
```

delawski commented on src/Admin/SupportScreen.php:
```diff
@@ -182,16 +182,44 @@ public function enqueue_assets( $hook_suffix ) {
+
+		return [
+			'all'   => intval( $amp_validated_post_count->publish ),
+			'valid' => intval( $query->found_posts ),
```
Would `fresh` be a more accurate name for this? Or conversely, how about inverting the logic on the frontend and returning here `stale` count instead of `valid`/`fresh`?

delawski commented on src/Support/SupportData.php:
```diff
@@ -620,11 +652,98 @@ static function ( $url ) {
+				}
+
+				// If validation data is exists for URL then check if those data stalled or not, If stalled then revalidate.
+				$is_stalled = ! empty( AMP_Validated_URL_Post_Type::get_post_staleness( $amp_validated_post_id ) );
+				if ( $is_stalled ) {
```
```suggestion
				$is_stale = ! empty( AMP_Validated_URL_Post_Type::get_post_staleness( $amp_validated_post_id ) );
				if ( $is_stale ) {
```

delawski reviewed (changes requested):
I've reviewed and tested this PR locally. Everything seems to be working as expected for me.

I left some feedback on the frontend part and briefly browsed the backend code. @bartoszgadomski would you be able to do a more thorough code review (mainly the backend part)?

bartoszgadomski commented on src/Admin/SupportLink.php:
```diff
@@ -112,8 +110,8 @@ public function amp_validated_url_status_actions( $actions, WP_Post $post ) {
-			'page'    => 'amp-support',
-			'post_id' => $post->ID,
+			'page' => 'amp-support',
+			'url'  => rawurlencode( $post->post_title ),
```
@dhaval-parekh This will not generate post url:

![Screenshot 2021-11-22 at 10 34 31](https://user-images.githubusercontent.com/4476915/142837351-3d85b24c-dabf-4017-a518-00f545fb0560.png)

Probably something like this can be used instead:

```suggestion
			'url'  => get_permalink( $post ),
```

dhaval-parekh replied:
@bartoszgadomski That would be great if you provide steps to reproduce.

We need to use `$post->post_title` since here we need a URL that is validated instead of the URL of that post itself. And we are storing validated URL as post title  (in `amp_validated_url` post type)

bartoszgadomski replied:
I see... I tried to broke your code and looks like I went too far with that. As this piece of code is dedicated for `amp_validated_url` post type only, I think we are ok 👍

bartoszgadomski commented on src/Admin/SupportLink.php:
```diff
@@ -140,8 +138,8 @@ public function post_row_actions( $actions, WP_Post $post ) {
-			'page'    => 'amp-support',
-			'post_id' => $post->ID,
+			'page' => 'amp-support',
+			'url'  => rawurlencode( $post->post_title ),
```
@dhaval-parekh Same as in https://github.com/ampproject/amp-wp/pull/6718#discussion_r754094354

westonruter replied:
```suggestion
			'url'  => rawurlencode( AMP_Validated_URL_Post_Type::get_url_from_post( $post ) ),
```

bartoszgadomski commented on src/Admin/SupportScreen.php:
```diff
@@ -163,13 +165,11 @@ public function enqueue_assets( $hook_suffix ) {
+		$amp_url = isset( $_GET['url'] ) ? esc_url_raw( $_GET['url'] ) : ''; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
 
-		if ( ! empty( $post_id ) && 0 < intval( $post_id ) ) {
+		if ( ! empty( $amp_url ) ) {
```
@dhaval-parekh Maybe we could validate here if a given string is actually an url? Something like:

```suggestion
		if ( ! empty( $amp_url ) && filter_var( $amp_url, \FILTER_VALIDATE_URL ) ) {
```

bartoszgadomski commented on src/Support/SupportData.php:
```diff
@@ -551,18 +557,33 @@ static function ( $plugin ) {
 		 */
-		if ( empty( $source['type'] ) || ! in_array( $source['type'], $allowed_types, true ) ) {
+		if (
+			( empty( $source['type'] ) || ! in_array( $source['type'], $allowed_types, true ) ) &&
```
@dhaval-parekh Maybe move the `&&` into new line to improve readability:

```suggestion
			( empty( $source['type'] ) || ! in_array( $source['type'], $allowed_types, true ) )
			&&
			...
```

bartoszgadomski commented on src/Support/SupportData.php:
```diff
@@ -620,11 +652,98 @@ static function ( $url ) {
+			// If no specific URL requested.
+			$query_args = [
+				'post_type'      => AMP_Validated_URL_Post_Type::POST_TYPE_SLUG,
+				'posts_per_page' => 100,
```
@dhaval-parekh Why there's a posts limit set to 100?

dhaval-parekh replied:
It's because there can be a lot of `amp_validated_post` posts and we shouldn't be fetching all those but also need to fetch a fair amount of data so the user can send it.

bartoszgadomski commented on src/Support/SupportData.php:
```diff
@@ -620,11 +652,98 @@ static function ( $url ) {
+
+				if ( empty( $amp_validated_post_id ) ) {
+					$validity              = AMP_Validation_Manager::validate_url_and_store( $url );
+					$amp_validated_post_id = $validity['post_id'];
```
@dhaval-parekh As I mentioned in one of the comments above (see https://github.com/ampproject/amp-wp/pull/6718#discussion_r754094354), the "url" attribute made of post title ends up with something like `"https://Test%20embeds"` here, which then throws a Fatal error:

`Fatal error: Uncaught Error: Cannot use object of type WP_Error as array in /var/www/html/wp-content/plugins/amp-wp/src/Support/SupportData.php:692 Stack trace: #0 /var/www/html/wp-content/plugins/amp-wp/src/Support/SupportData.php(195): AmpProject\AmpWP\Support\SupportData->get_amp_urls() #1 /var/www/html/wp-content/plugins/amp-wp/src/Admin/SupportScreen.php(177): AmpProject\AmpWP\Support\SupportData->get_data() #2 /var/www/html/wp-includes/class-wp-hook.php(303): AmpProject\AmpWP\Admin\SupportScreen->enqueue_assets('amp_page_amp-su...') #3 /var/www/html/wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters(NULL, Array) #4 /var/www/html/wp-includes/plugin.php(470): WP_Hook->do_action(Array) #5 /var/www/html/wp-admin/admin-header.php(102): do_action('admin_enqueue_s...', 'amp_page_amp-su...') #6 /var/www/html/wp-admin/admin.php(239): require_once('/var/www/html/w...') #7 {main} thrown in /var/www/html/wp-content/plugins/amp-wp/src/Support/SupportData.php on line 692`

dhaval-parekh replied:
I have added a check for it to handle that scenario.

dhaval-parekh committed:
Update AMP support notice component

dhaval-parekh committed:
Validate URL before passing to SupportData service

dhaval-parekh committed:
Apply suggestion from PR

dhaval-parekh commented:
While testing this PR. Found a bug that when there are more than 100 `amp_validated_url` posts and the first 100 doesn't have any error. Support page display notices that "Site doesn't have AMP validation errors". But there are validation errors and the user can see errors in the "Validated URL" Page.

This can be fixed quickly by adding ` 's' => 'term_slug' ` in query args to fetch only `amp_validated_url` posts that have AMP validation errors. Or looping through all the `amp_validated_url`. [here](https://github.com/ampproject/amp-wp/pull/6718/files/fb5bb2702b76e875483018d576c5b0a2471310ee#diff-c72c06963f4244e78c244f5ffc9351ce1bbd67b0a3af998bff5c585c6b55f15aR657)

However, the problem with the first approach is it might skip the posts that have a CSS budget of more than 100%.

dhaval-parekh committed:
Update prop types

dhaval-parekh committed:
Fetch only amp_validated_url post that have error on support page

delawski commented on assets/src/components/amp-support/validation-results-notice.js:
```diff
@@ -0,0 +1,101 @@
+export function ValidationResultsNotice( { data, args, ampValidatedPostCount } ) {
+	const isSpecificUrlRequest = 0 < args?.urls?.length;
+	const hasErrors = 0 < data.errors?.length || 0 < data.urls?.length;
+	const hasAllStaleResult = 0 < ampValidatedPostCount.all && ampValidatedPostCount.all === ampValidatedPostCount.stale;
+	const hasSomeStaleResult = 0 < ampValidatedPostCount.fresh && 0 < ampValidatedPostCount.stale;
+	const hasAllFreshResult = 0 < ampValidatedPostCount.all && ampValidatedPostCount.all === ampValidatedPostCount.fresh;
```
NIT: How about using plural nouns in the constant names, i.e. `hasAllStaleResults`, `hasSomeStaleResults` and `hasAllFreshResult`?

dhaval-parekh replied:
I have addressed this in [c98d9be](https://github.com/ampproject/amp-wp/pull/6718/commits/c98d9be81a02c762c1950e2da6876592b91bf3af)

delawski reviewed (commented):
Thank you for addressing the code review feedback. It looks good to me now 👍

dhaval-parekh committed:
Use plural nouns in variable name

dhaval-parekh committed:
Fix typo in Sandboxinf level drawer

bartoszgadomski reviewed (approved):
Looks good to me 👍

dhaval-parekh committed:
Merge branch 'develop' into bug/6707-support-request

dhaval-parekh committed:
Merge branch 'develop' into bug/6707-support-request

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into bug/6707-support-request

* 'develop' of github.com:ampproject/amp-wp: (39 commits)
  Harden is_needed check to include has_cap
  Fix another usage of Encoding
  Sort imports for impacted files
  Add noopener to external link
  Update to amp-toolbox 0.9.1
  Open validated URLs in new window to prevent accidental loss
  Show URLs with validation issues to non-technical users as well
  Cover case when Gutenberg is not active with E2E tests
  Do not scan site for network activated plugins
  Enqueue missing `wp-components` CSS dependency
  Ensure timeout has been set before trying to clear it
  Update E2E tests
  Improve plugin activation post-site-scan message copy
  Ensure demo plugin is deactivated before running tests
  Update plugin activation message if validation errors are found
  Update E2E tests to reflect recent changes
  Always pass 2 arguments to `visitAdminPage()`
  Rename `AdminNotice` component to `AmpAdminNotice`
  Cleanly skip test if dependency is missing
  Apply notice text update suggestions from code review
  ...

westonruter commented on assets/src/components/amp-support/validation-results-notice.js:
```diff
@@ -0,0 +1,101 @@
+		);
+	}
+
+	// The site doesn't have any AMP errors but there are some stalled result.
```
```suggestion
	// The site doesn't have any AMP errors but there are some stale results.
```

westonruter commented on src/Admin/SupportScreen.php:
```diff
@@ -182,16 +183,45 @@ public function enqueue_assets( $hook_suffix ) {
+	 *
+	 * @return array [
+	 *     @type int $all Count of all AMP validated URL post.
+	 *     @type int $valid Count of all AMP validated URL post that are not stalled.
```
```suggestion
	 *     @type int $valid Count of all AMP validated URL posts that are not stale.
```

westonruter commented on src/Support/SupportData.php:
```diff
@@ -620,11 +658,104 @@ static function ( $url ) {
+					$amp_validated_post_id = $validity['post_id'];
+				}
+
+				// If validation data is exists for URL then check if those data stalled or not, If stalled then revalidate.
```
```suggestion
				// If validation data is exists for URL then check if it is stale or not. If it is stale, then revalidate.
```

westonruter commented on src/Support/SupportData.php:
```diff
@@ -620,11 +658,104 @@ static function ( $url ) {
+				// If validation data is exists for URL then check if those data stalled or not, If stalled then revalidate.
+				$is_stale = ! empty( AMP_Validated_URL_Post_Type::get_post_staleness( $amp_validated_post_id ) );
+				if ( $is_stale ) {
+					AMP_Validation_Manager::validate_url_and_store( $url );
```
```suggestion
					AMP_Validation_Manager::validate_url_and_store( $url, $amp_validated_post_id );
```

westonruter commented on src/Support/SupportData.php:
```diff
@@ -620,11 +658,104 @@ static function ( $url ) {
+			foreach ( $this->urls as $url ) {
+
+				// Check request URL have validated data if not then validate URL.
+				$amp_validated_post_id = post_exists( $url, '', '', AMP_Validated_URL_Post_Type::POST_TYPE_SLUG );
```
This should rather use `AMP_Validated_URL_Post_Type::get_invalid_url_post( $url )`

westonruter commented on src/Admin/SupportScreen.php:
```diff
@@ -182,16 +183,45 @@ public function enqueue_assets( $hook_suffix ) {
+	 *     @type int $valid Count of all AMP validated URL post that are not stalled.
+	 * ]
+	 */
+	public function get_amp_validated_post_counts() {
```
Can tests be added for this?

dhaval-parekh replied:
Added test cases in [575dba3](https://github.com/ampproject/amp-wp/pull/6718/commits/575dba32094e0ef705cb0b3c03b8fbc728821c63)

westonruter commented on src/Admin/SupportLink.php:
```diff
@@ -112,8 +110,8 @@ public function amp_validated_url_status_actions( $actions, WP_Post $post ) {
-			'page'    => 'amp-support',
-			'post_id' => $post->ID,
+			'page' => 'amp-support',
+			'url'  => rawurlencode( $post->post_title ),
```
It would be preferable to use `\AMP_Validated_URL_Post_Type::get_url_from_post()` to obtain the URL from a validated URL post:

```suggestion
			'url'  => rawurlencode( AMP_Validated_URL_Post_Type::get_url_from_post( $post ) ),
```

westonruter commented on src/Admin/SupportScreen.php:
```diff
@@ -162,14 +164,13 @@ public function enqueue_assets( $hook_suffix ) {
-				'amp_validated_post_ids' => [
-					$post_id,
-				],
+				'urls' => [ esc_url_raw( $amp_url ) ],
```
```suggestion
				'urls' => [ $amp_url ],
```

westonruter commented on src/Admin/SupportScreen.php:
```diff
@@ -162,14 +164,13 @@ public function enqueue_assets( $hook_suffix ) {
-		$args    = [];
-		$post_id = isset( $_GET['post_id'] ) ? (int) $_GET['post_id'] : 0; // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		$args = [];
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended, WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+		$amp_url = isset( $_GET['url'] ) ? $_GET['url'] : '';
```
```suggestion
		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
		$amp_url = isset( $_GET['url'] ) ? esc_url_raw( wp_unslash( $_GET['url'] ) ) : '';
```

westonruter commented on src/Support/SupportData.php:
```diff
@@ -620,11 +658,104 @@ static function ( $url ) {
+		}
+
+		// Empty array for post staleness means post is NOT stale.
+		if ( ! empty( AMP_Validated_URL_Post_Type::get_post_staleness( $amp_validated_post->ID ) ) ) {
```
The method takes a `WP_Post` object as well:

```suggestion
		if ( ! empty( AMP_Validated_URL_Post_Type::get_post_staleness( $amp_validated_post ) ) ) {
```

westonruter reviewed (changes requested):
I'm testing this but it's not working for me. I'm getting an infinite series of HTTP requests that end in 500 errors:


https://user-images.githubusercontent.com/134745/144539989-7483f3d3-097d-4744-85a6-1a9952b398ac.mov

The console shows:

<img width="595" alt="Screen Shot 2021-12-02 at 19 13 37" src="https://user-images.githubusercontent.com/134745/144540015-1feddf84-272a-4978-b134-8d51885223d2.png">

Error log has:

````
[03-Dec-2021 03:13:54 UTC] PHP Fatal error:  Uncaught Error: Call to undefined function AmpProject\AmpWP\Support\post_exists() in /app/public/content/plugins/amp/src/Support/SupportData.php:695
Stack trace:
#0 /app/public/content/plugins/amp/src/Support/SupportData.php(200): AmpProject\AmpWP\Support\SupportData->get_amp_urls()
#1 /app/public/content/plugins/amp/src/Support/SupportData.php(159): AmpProject\AmpWP\Support\SupportData->get_data()
#2 /app/public/content/plugins/amp/src/Support/SupportRESTController.php(107): AmpProject\AmpWP\Support\SupportData->send_data()
#3 /app/public/core-dev/src/wp-includes/rest-api/class-wp-rest-server.php(1141): AmpProject\AmpWP\Support\SupportRESTController->callback(Object(WP_REST_Request))
#4 /app/public/core-dev/src/wp-includes/rest-api/class-wp-rest-server.php(988): WP_REST_Server->respond_to_request(Object(WP_REST_Request), '/amp/v1/send-di...', Array, NULL)
#5 /app/public/core-dev/src/wp-includes/rest-api/class-wp-rest-server.php(414): WP_REST_Server->dispatch(Object(WP_REST_Request))
#6 /app/public/core-dev/s in /app/public/content/plugins/amp/src/Support/SupportData.php on line 695
```

In the case of an error, it shouldn't automatically try to re-send. It should prompt the user to push the submit button again.

dhaval-parekh committed:
Update function docs.
Use AMP_Validated_URL_Post_Type::get_invalid_url_post() instead of post_exists()

dhaval-parekh committed:
Add test case for SupportScreen::get_amp_validated_post_counts()

dhaval-parekh committed:
Prevent multiple support request in case of error

dhaval-parekh committed:
Apply suggestion from PR

dhaval-parekh committed:
Sanitize value of query param before using

dhaval-parekh commented:
> In the case of an error, it shouldn't automatically try to re-send. It should prompt the user to push the submit button again.

I have updated it. And now in case of error, it will show an error message to the user.

---

Regarding error. While checking further found that `post_exists()` is available in admin requests and WP CLI but not in REST endpoint which was causing the issue during sending a support request.

Now instead of using `post_exists()`  now using `AMP_Validated_URL_Post_Type::get_invalid_url_post()` that will resolve the issue.

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into bug/6707-support-request

* 'develop' of github.com:ampproject/amp-wp:
  Clear timeout correctly
  Simplify the E2E test
  Simplify logic for ensuring consistent units
  Propagate more styles AMP-valid dimensions to AMP element
  Add tests for deriving width/height from styles
  Propagate width/height styles as width/height attributes when setting layout

westonruter commented on src/Support/SupportRESTController.php:
```diff
@@ -67,6 +67,19 @@ public static function get_registration_action() {
 	 */
 	public function register() {
 
+		register_rest_route(
+			$this->namespace,
+			'/get-diagnostic',
+			[
+				[
+					'methods'             => WP_REST_Server::READABLE,
+					'args'                => [],
+					'permission_callback' => [ $this, 'permission_callback' ],
+					'callback'            => [ $this, 'get_item' ],
+				],
+			]
+		);
+
```
This endpoint isn't being used anywhere, and it's not validating the `args`. So I think it should be removed.

westonruter committed:
Further reuse amp_validated_post variable

westonruter committed:
Remove unused REST API endpoint

Partially revert "Update function docs."

This reverts commit 262371d54fe6418ecbb184575a7c794135778994.

westonruter commented:
I'm still getting an error when submitting this, but this time it seems to be an issue with the insights server. It's responding with:

```json
{
    "status": "fail",
    "data": [
        {
            "message": "required validation failed on plugins.0.version",
            "field": "plugins.0.version",
            "validation": "required"
        },
        {
            "message": "required validation failed on plugins.5.version",
            "field": "plugins.5.version",
            "validation": "required"
        },
        {
            "message": "required validation failed on plugins.6.version",
            "field": "plugins.6.version",
            "validation": "required"
        }
    ]
}
```

It is because I have active plugins which do not have a `version`. This should not be a required field.

Filed as https://github.com/rtCamp/px-wp-insights/issues/186.

westonruter committed:
Translate error message when failing to send to insights server

westonruter commented on src/Admin/SupportScreen.php:
```diff
@@ -182,16 +183,45 @@ public function enqueue_assets( $hook_suffix ) {
+			'posts_per_page' => 1,
+			'post_status'    => 'publish',
+			'meta_key'       => AMP_Validated_URL_Post_Type::VALIDATED_ENVIRONMENT_POST_META_KEY,
+			'meta_value'     => maybe_serialize( AMP_Validated_URL_Post_Type::get_validated_environment() ), // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value
```
This is a clever way to query for non-stale results.

westonruter commented on src/Support/SupportData.php:
```diff
@@ -636,126 +766,106 @@ static function ( $url ) {
 		}
 
+		$amp_invalid_url = [
+			'url'                   => $amp_validated_post->post_title,
```
This and other instances in the file could use `\AMP_Validated_URL_Post_Type::get_url_from_post()`.

dhaval-parekh replied:
Updated in new PR #6764

westonruter commented on src/Support/SupportData.php:
```diff
@@ -620,11 +658,103 @@ static function ( $url ) {
+				'post_status'    => 'publish',
+				'meta_key'       => AMP_Validated_URL_Post_Type::VALIDATED_ENVIRONMENT_POST_META_KEY,
+				'meta_value'     => maybe_serialize( AMP_Validated_URL_Post_Type::get_validated_environment() ), // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value
+				's'              => 'term_slug',
```
Why search for `term_slug`? Just something that exists in the `post_content`?

dhaval-parekh replied:
Added `term_slug` as a search query string to make sure any data we fetch has an error and is not empty. 
`term_slug` will always be available in post content for `amp_invalidated_url` posts that have errors.

westonruter commented on src/Support/SupportData.php:
```diff
@@ -551,18 +562,34 @@ static function ( $plugin ) {
 		$allowed_types  = [ 'plugin', 'theme' ];
 		$source['type'] = ( ! empty( $source['type'] ) ) ? strtolower( trim( $source['type'] ) ) : '';
 
+		if ( ! empty( $source['sources'] ) && is_array( $source['sources'] ) ) {
+			foreach ( $source['sources'] as $index => $inner_source ) {
+				$source['sources'][ $index ] = self::normalize_error_source( $inner_source );
+			}
+
+			$source['sources'] = array_values( array_filter( $source['sources'] ) );
+		}
```
This was added in 9c8ef66716264bef2290af9ff9f0ecc1bc54684a to add normalization of nested sources?

dhaval-parekh replied:
Yes, I have added to normalize nested sources.

westonruter committed:
Improve phpdoc comments and guard against JSON parse error for post_content

westonruter committed:
Fix test after message update in 3a5f4aa9
